### PR TITLE
[SelectionDAG] Add m_Neg and m_Not pattern matcher and update DAGCombiner

### LINKED ISSUE #85065: [DAG] SDPatternMatch - add m_Neg and m_Not matchers
I've started using the SDPatternMatch.h matchers to simplify dag combines (especially commutable ones - thank you for that!).

We're currently missing a couple of convenience matchers for
```c
m_Neg(m_Value(X)) == m_Sub(m_Zero(), m_Value(X))
m_Not(m_Value(X)) == m_Xor(m_Value(X), m_AllOnes())
```
You will need some experience with C++ templates to understand the pattern matching system to add these.

CC @mshockwave

### PR DESCRIPTION
**Pull request by ZiCong-Wang**
Resolves #85065

### CODE REVIEW

File: llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp [hunk 1]
```diff
@@ -174,10 +174,18 @@ TEST_F(SelectionDAGPatternMatchTest, matchUnaryOp) {
   SDValue SExt = DAG->getNode(ISD::SIGN_EXTEND, DL, Int64VT, Op0);
   SDValue Trunc = DAG->getNode(ISD::TRUNCATE, DL, Int32VT, Op1);
 
+  SDValue Zero = DAG->getConstant(0, DL, Int32VT);
+  SDValue AllOnes = DAG->getConstant(APInt::getAllOnes(32), DL, Int32VT);
+  SDValue Neg = DAG->getNode(ISD::SUB, DL, Int32VT, Zero, Op0);
+  SDValue Not = DAG->getNode(ISD::XOR, DL, Int32VT, AllOnes, Op0);
```
Comment (arsenm) [inline llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp:180]: There is already a DAG->getNOT helper for this

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 2]
```diff
@@ -705,6 +705,20 @@ inline auto m_False() {
...
 }
+
+/// Match a negtive as a sub(0, v)
+template <typename ValTy>
+inline BinaryOpc_match<SpecificInt_match, ValTy>
+m_Neg(const ValTy &V) {
+  return m_Sub(m_Zero(), V);
+}
+
+/// Match a Not as a xor(v, 1) or xor(1, v)
```
Comment (jayfoad) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:716]:
```suggestion
/// Match a Not as a xor(v, -1) or xor(-1, v)
```
Or `~0`

File: llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp (hunk 1 above)
Comment (ZiCong-Wang) [reply llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp:180]: Got it. The same to DAG->getNegative.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 3]
```diff
@@ -705,6 +705,20 @@ inline auto m_False() {
...
+template <typename ValTy>
+inline BinaryOpc_match<SpecificInt_match, ValTy, true>
+m_Not(const ValTy &V) {
+  return m_Xor(m_AllOnes(), V);
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:720]: This should try matching the constant in the RHS operand, as that's canonical.

The IR pattern matchers have separate _c versions for versions that try to commute the inputs

**Review by arsenm — APPROVED** [review]

File: llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp [hunk 4]
```diff
@@ -174,10 +174,19 @@ TEST_F(SelectionDAGPatternMatchTest, matchUnaryOp) {
...
+  SDValue Not = DAG->getNOT(DL, Op0, Int32VT);
+
   using namespace SDPatternMatch;
   EXPECT_TRUE(sd_match(ZExt, m_UnaryOp(ISD::ZERO_EXTEND, m_Value())));
   EXPECT_TRUE(sd_match(SExt, m_SExt(m_Value())));
   EXPECT_TRUE(sd_match(Trunc, m_Trunc(m_Specific(Op1))));
+
+  EXPECT_TRUE(sd_match(Neg, m_Neg(m_Value())));
+  EXPECT_TRUE(sd_match(Not, m_Not(m_Value())));
+  EXPECT_FALSE(sd_match(Neg, m_Node(ISD::SUB, m_Value())));
```
Comment (mshockwave) [inline llvm/unittests/CodeGen/SelectionDAGPatternMatchTest.cpp:187]: I think what Matt meant was EXPECT_FALSE test against your new m_Neg and m_Not. For instance, check if `sd_match(ZExt, m_Neg(...))` is false.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 5]
```diff
@@ -705,6 +705,18 @@ inline auto m_False() {
       },
       m_Value()};
 }
+
+/// Match a negtive as a sub(0, v)
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:709]:
```suggestion
/// Match a negate as a sub(0, v)
```

**Pull request merged**